blockchain/v2/reactor_test.go

```diff
@@ -1,22 +1,552 @@
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			reactor := newTestReactor(params)
+			reactor.Start()
+			mockSwitch := &mockSwitchIo{switchedToConsensus: false}
+			reactor.io = mockSwitch
+			// time for go routines to start
+			time.Sleep(time.Millisecond)
+
+			for i := 0; i < len(tt.msgs); i++ {
+				step := tt.msgs[i]
+				switch ev := step.event.(type) {
+				case bcStatusRequestMessage:
+					old := mockSwitch.numStatusResponse
+					reactor.Receive(channelID, mockPeer{id: p2p.ID(step.peer)}, cdc.MustMarshalBinaryBare(ev))
+					assert.Equal(t, old+1, mockSwitch.numStatusResponse)
+				case bcBlockRequestMessage:
+					if ev.Height > params.startHeight {
+						old := mockSwitch.numNoBlockResponse
+						reactor.Receive(channelID, mockPeer{id: p2p.ID(step.peer)}, cdc.MustMarshalBinaryBare(ev))
+						assert.Equal(t, old+1, mockSwitch.numNoBlockResponse)
+					} else {
+						old := mockSwitch.numBlockResponse
+						reactor.Receive(channelID, mockPeer{id: p2p.ID(step.peer)}, cdc.MustMarshalBinaryBare(ev))
+						assert.Equal(t, old+1, mockSwitch.numBlockResponse)
+					}
+				}
+			}
+			time.Sleep(time.Millisecond)
+
+			reactor.Stop()
+		})
+	}
+}
+
+//----------------------------------------------
+// utility funcs
+
+func makeTxs(height int64) (txs []types.Tx) {
+	for i := 0; i < 10; i++ {
+		txs = append(txs, types.Tx([]byte{byte(height), byte(i)}))
+	}
+	return txs
+}
+
+func makeBlock(height int64, state sm.State, lastCommit *types.Commit) *types.Block {
+	block, _ := state.MakeBlock(height, makeTxs(height), lastCommit, nil, state.Validators.GetProposer().Address)
+	return block
+}
+
+type testApp struct {
+	abci.BaseApplication
+}
+
+func makeVote(header *types.Header, blockID types.BlockID, valset *types.ValidatorSet, privVal types.PrivValidator) *types.Vote {
+	addr := privVal.GetPubKey().Address()
+	idx, _ := valset.GetByAddress(addr)
+	vote := &types.Vote{
+		ValidatorAddress: addr,
+		ValidatorIndex:   idx,
+		Height:           header.Height,
+		Round:            1,
+		Timestamp:        tmtime.Now(),
+		Type:             types.PrecommitType,
+		BlockID:          blockID,
+	}
+
+	_ = privVal.SignVote(header.ChainID, vote)
+
+	return vote
+}
+
+func randGenesisDoc(chainID string, numValidators int, randPower bool, minPower int64) (*types.GenesisDoc, []types.PrivValidator) {
+	validators := make([]types.GenesisValidator, numValidators)
+	privValidators := make([]types.PrivValidator, numValidators)
+	for i := 0; i < numValidators; i++ {
+		val, privVal := types.RandValidator(randPower, minPower)
+		validators[i] = types.GenesisValidator{
+			PubKey: val.PubKey,
+			Power:  val.VotingPower,
+		}
+		privValidators[i] = privVal
+	}
+	sort.Sort(types.PrivValidatorsByAddress(privValidators))
+
+	return &types.GenesisDoc{
+		GenesisTime: tmtime.Now(),
+		ChainID:     chainID,
+		Validators:  validators,
+	}, privValidators
+}
+
+func newReactorStore(
```

Comment:
This function requires a rather large dependency that extends beyond the boundaries of this package. These dependencies would then become a liability for any refactoring happening in those external packages. The blockchain reactor `NewReactor` constructor depends only on an interface that should be mockable. Are we certain that we need a real store.BlockStore here?